File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <iostream>
#include <vector>
using namespace std;
int r,c,maxlen=0;
vector<vector<char>> arr;
vector<int> visit(26,0);
int mx[4] ={0,0,1,-1};
int my[4] ={1,-1,0,0};

void dfs(int x, int y, int sum) {
    int a = arr[x][y]-'A';
    visit[a]=1;
    maxlen = max(maxlen, sum);
    
    for(int i=0; i<4; i++) {
        int nx = x+mx[i];
        int ny = y+my[i];
        int nn = arr[x+mx[i]][y+my[i]] - 'A';
        if(nx>=0 && nx<r && ny>=0 && ny<c) {
            if(visit[nn]==0) {
                dfs(x+mx[i],y+my[i],sum+1);
                visit[nn]=0;
            }
        }
    }
}

int main() {
    cin>>r>>c;
    arr.resize(r,vector<char>(c,0));
    for(int i=0; i<r; i++) {
        for(int j=0; j<c; j++) {
            cin>>arr[i][j];
        }
    }
    dfs(0,0,1);
    
    cout<<maxlen<<'\n';
    return 0;
    
    
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

public class Main {
    static int r,c,maxLen;
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    static char[][] arr;
    static boolean[] visited = new boolean[26];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        arr = new char[r][c];

        for (int i = 0; i < r; i++) {
            String[] s = br.readLine().split("");
            for (int j = 0; j < c; j++) {
                arr[i][j] = s[j].charAt(0);
            }
        }
        dfs(0,0,1);
        System.out.println(maxLen);
    }

    public static void dfs(int i, int j, int depth) {
        int alpaIdx = arr[i][j] - 'A';
        visited[alpaIdx] = true;
        maxLen = Math.max(maxLen, depth);

        for (int k = 0; k < 4; k++) {
            int nx = i + dx[k];
            int ny = j + dy[k];

            if(nx>=0 && nx<r && ny>=0 && ny<c) {
                int nextAlpaIdx = arr[nx][ny] - 'A';
                if(!visited[nextAlpaIdx]) {
                    dfs(nx, ny, depth+1);
                    visited[nextAlpaIdx] = false;
                }
            }
        }
    }
}